### PR TITLE
Add support for js-sources

### PR DESCRIPTION
**Pull request by hsyl20**
This PR adds support for `js-sources` with GHC, not only GHCJS.

Caveat: for now, similarly to `asm-sources` and `cmm-sources`, `js-sources` only works for `library` components, not `executable` components. See https://github.com/haskell/cabal/issues/8639

Not sure how to add tests for this as it requires GHC >= 9.6 built for the JS target. I've only tested it locally.

---

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

### CODE REVIEW

**Comment by hsyl20** [comment]
I've opened https://github.com/haskell/cabal/issues/8639 to track the issue with executable components. It has a larger scope because `asm-sources` and `cmm-sources` are concerned too.

**Comment by Mikolaj** [comment]
Local tests are fine in this case. A tiny changelog file would be very welcome though.

@hsyl20: I presume this is ready for review? Based on this assumption, I've set the label and asked @bgamari to have a look (or delegate to somebody in GHC team that works on related issues).

Re including in GHC 9.6, we plan to release cabal 3.10 in step with GHC 9.6, so if this PR gets merged in 2022 or shortly thereafter, it's going to be included in GHC 9.6 and in cabal 3.10. I will bug people to make sure it doesn't just get forgotten.

**Comment by hsyl20** [comment]
I've added a changelog. Now it's ready for review. :)

File: Cabal/src/Distribution/Simple/GHC.hs
Comment (Mikolaj) [inline Cabal/src/Distribution/Simple/GHC.hs:564]: This is adding JS to `cLikeSources`. Is "C-like" a misnomer or was the original intention to exclude JS? Change of plans? I don't see any "non-C-like sources" list in the codebase that we could add JS source to, instead, so this must be the right place. 

Perhaps the C-like vs non-C-like distinction is a left over from some special cases for GHCJS? Is JS C-like enough when considered by GHC JS backend, but not C-like when fed to GHCJS?

Comment (hsyl20) [reply Cabal/src/Distribution/Simple/GHC.hs:564]: Yes for the JS backend we've reworked the way JS sources are treated. They are now C-like because we've added support for a "compilation" from `.js` to `.o` files (basically a file renaming with a header added). It makes them less special for tools like Hadrian and cabal.

(in the GHCJS code path (`Distribution.Simple.GHCJS`) these files were directly passed as link options)

**Review by Mikolaj — APPROVED** [review]
I've verified that this PR changes all the bits of code that are likely to need changing. However, I don't know this functionality enough to understand whether these changes makes sense, so I have to trust the results of local testing by the PR author. This is unlikely to break other stuff, though.

File: Cabal/src/Distribution/Simple/GHC.hs
Comment (Profpatsch) [reply Cabal/src/Distribution/Simple/GHC.hs:564]: Maybe this deserves a `NOTE JS Sources are C-Like` or similar.

Comment (Mikolaj) [reply Cabal/src/Distribution/Simple/GHC.hs:564]: @Profpatsch: GHC note or cabal note? Or one of those and hyperlink from the other place?

Comment (hsyl20) [reply Cabal/src/Distribution/Simple/GHC.hs:564]: I've added a comment explaining that JS sources are C-like in the JS backend.

**Comment by jneira** [comment]
hi, this looks great but it would be nice to know the plans to add or modify existing tests

**Comment by jneira** [comment]
> Not sure how to add tests for this as it requires GHC >= 9.6 built for the JS target. I've only tested it locally.

A possible way would be write the tests and run them conditionally on ghc version. They will be eventually executed everywhere.

**Comment by jneira** [comment]
This pr seems to be related with #8633 and i guess tests also would be related

**Comment by hsyl20** [comment]
> A possible way would be write tests and run them conditionally on ghc version. They will be eventually executed everywhere.

I'll try to do this.

> This pr seems to be related with https://github.com/haskell/cabal/pull/8633 and i guess tests also would be related

The two PRs aren't related. This other one is for the wasm backend, this one is for the JS backend.

**Comment by Mikolaj** [comment]
We'd like to include this commit in the 3.10 branch that is to be cut RSN so, in the worst case, perhaps we could add the test afterwards. However, we need a second review, an in-depth one, ideally prompting some more documentation to be added, so I'm kindly asking @ben to help with that.

**Comment by Mikolaj** [comment]
BTW, if this 

https://github.com/haskell/cabal/issues/8675

is done in finite time and we have 9.6.1-alpha1 in CI, the test would be able to run. Probably not on Windows, though, because our CI is quite thoroughly broken with GHC >= 9.4 on Windows.

**Comment by hsyl20** [comment]
@Mikolaj You also will need a build of GHC cross-compiling for the JS target. Not sure if it is planned?

**Comment by hsyl20** [comment]
I've added a test but I couldn't run it with `cabal-tests`. It doesn't find the correct `ghc-pkg` and there is no `--with-ghc-pkg` flag use.

**Comment by hsyl20** [comment]
Oops I forgot to push some changes, now done. Run example:

```
/home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.4/cabal-install-3.9.0.0/x/cabal/build/cabal/cabal --with-ghc='/home/hsyl20/projects/ghc/javascript-backend/_build/stage1/bin/js-unknown-ghcjs-ghc' --with-ghc-pkg='/home/hsyl20/projects/ghc/javascript-backend/_build/stage1/bin/js-unknown-ghcjs-ghc-pkg' run demo
Warning: this is a debug build of cabal-install with assertions enabled.
Warning: Unknown/unsupported 'ghc' version detected (Cabal 3.9.0.0 supports
'ghc' version < 9.6):
/home/hsyl20/projects/ghc/javascript-backend/_build/stage1/bin/js-unknown-ghcjs-ghc
is version 9.7.20230112
Resolving dependencies...
Build profile: -w ghc-9.7.20230112 -O1
In order, the following will be built (use -v for more details):
 - jssources-0 (lib) (first run)
 - jssources-0 (exe:demo) (first run)
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring library for jssources-0..
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for jssources-0..
Building library for jssources-0..
[1 of 1] Compiling Lib              ( src/Lib.hs, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/js-ghcjs/ghc-9.7.20230112/jssources-0/
Warning: this is a debug build of cabal-install with assertions enabled.
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring executable 'demo' for jssources-0..
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing executable 'demo' for jssources-0..
Building executable 'demo' for jssources-0..
[1 of 1] Compiling Main             ( demo/Main.hs, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/js-ghcjs/ghc-9.7.20230112/jssources-0/x/demo/build/demo/demo-tmp/Main.o )
[2 of 2] Linking /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/js-ghcjs/ghc-9.7.20230112/jssources-0/x/demo/build/demo/demo.jsexe
Hello JS!
```

File: cabal-testsuite/PackageTests/JS/JsSources/cabal.test.hs
Comment (Mikolaj) [inline cabal-testsuite/PackageTests/JS/JsSources/cabal.test.hs:5]: `skipUnless` needs a short String blurb as its first argument.

**Review by Mikolaj — CHANGES_REQUESTED** [review]
Thanks a lot for the test. I think it's fine as the first test of any functionality in any way related to JS.

**Comment by hsyl20** [comment]
I've also fixed `skipIfGhcVersion` in passing. It probably shouldn't belong to this MR but time is short.

**Review by Mikolaj — APPROVED** [review]
> I've also fixed skipIfGhcVersion in passing. It probably shouldn't belong to this MR but time is short.

Thanks a lot. It is. :)

**Comment by Mikolaj** [comment]
Looking at the excluding conditions in the test (and not seeing similar condition in the code proper), I realise I have no idea what happens if the new cabal with this PR is used together with GHC < 9.6. Does GHC fail gracefully? Does it just ignore the .js files? What if the old GHCJS is in play as well? Can it be used as before? Should we mentions in release notes (and docs) that cabal 3.10 requires GHC 9.6 for any JS compilation and is not compatible with the old GHCJS?

**Comment by hsyl20** [comment]
> I realise I have no idea what happens if the new cabal with this PR is used together with GHC < 9.6.

Good questions.

> Does GHC fail gracefully?

It will usually choke on the foreign imports like this:

```
❯ /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.4/cabal-install-3.9.0.0/x/cabal/build/cabal/cabal build
Warning: this is a debug build of cabal-install with assertions enabled.
Resolving dependencies...
Build profile: -w ghc-9.2.4 -O1
In order, the following will be built (use -v for more details):
 - jssources-0 (lib) (first run)
 - jssources-0 (exe:demo) (first run)
Configuring library for jssources-0..
Preprocessing library for jssources-0..
Building library for jssources-0..
[1 of 1] Compiling Lib              ( src/Lib.hs, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/x86_64-linux/ghc-9.2.4/jssources-0/build/Lib.o, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/x86_64-linux/ghc-9.2.4/jssources-0/build/Lib.dyn_o )

src/Lib.hs:3:1: error:
    • The `javascript' calling convention is unsupported on this platform
    • When checking declaration:
        foreign import javascript safe foo :: IO ()
  |
3 | foreign import javascript foo :: IO ()
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Same with 8.10, 9.0, and 9.4. However it doesn't do this with HEAD, so we've probably nuked the check by mistake... I'll open a ticket.

```
Building library for jssources-0..
[1 of 1] Compiling Lib              ( src/Lib.hs, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/x86_64-linux/ghc-9.7.20230113/jssources-0/build/Lib.o, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/x86_64-linux/ghc-9.7.20230113/jssources-0/build/Lib.dyn_o )
/usr/bin/ld.gold: error: cannot open dist-newstyle/build/x86_64-linux/ghc-9.7.20230113/jssources-0/build/jsbits/lib.dyn_o: No such file or directory
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
Error: cabal: Failed to build jssources-0 (which is required by exe:demo from
jssources-0).
```

> Does it just ignore the .js files?

If we remove the foreign imports, it returns the following link error with 8.10, 9.2, 9.4.

```
❯ /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.4/cabal-install-3.9.0.0/x/cabal/build/cabal/cabal build -w ghc-9.4
Warning: this is a debug build of cabal-install with assertions enabled.
Resolving dependencies...
Build profile: -w ghc-9.4.3 -O1
In order, the following will be built (use -v for more details):
 - jssources-0 (lib) (first run)
 - jssources-0 (exe:demo) (first run)
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for jssources-0..
Building library for jssources-0..
[1 of 1] Compiling Lib              ( src/Lib.hs, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/x86_64-linux/ghc-9.4.3/jssources-0/build/Lib.o, /home/hsyl20/projects/ghc/javascript-backend/libraries/Cabal/cabal-testsuite/PackageTests/JS/JsSources/dist-newstyle/build/x86_64-linux/ghc-9.4.3/jssources-0/build/Lib.dyn_o )
Warning: the following files would be used as linker inputs, but linking is not being done: jsbits/lib.js
ghc-9.4.3: no input files
Usage: For basic information, try the `--help' option.
Error: cabal: Failed to build jssources-0 (which is required by exe:demo from
jssources-0).
```

With HEAD we get the same link error as above.

> What if the old GHCJS is in play as well? Can it be used as before? Should we mentions in release notes (and docs) that cabal 3.10 requires GHC 9.6 for any JS compilation and is not compatible with the old GHCJS?

This PR doesn't change the codepath for GHCJS. So if GHCJS is detected as a compiler, it should work as before as far as I know.

The only change with this PR is that `js-sources` is supported with GHC, which is only useful starting with GHC 9.6. But people can already use old Cabal with GHC 9.6 with the JS backend. A workaround for `js-sources` not working is to pass JS files on the command line with `ghc-options: myjs.js`. It's just not very good UX :)

**Comment by hsyl20** [comment]
Ticket opened: https://gitlab.haskell.org/ghc/ghc/-/issues/22774

**Comment by hsyl20** [comment]
Second ticket: https://gitlab.haskell.org/ghc/ghc/-/issues/22775

**Comment by Mikolaj** [comment]
Splendid. That's descent behaviour and these are not blockers for this PR, IMHO.

**Review by bgamari — APPROVED** [review]

**Comment by Mikolaj** [comment]
@bgamari: thank you.

@hsyl20: Success! When you are ready, please kindly set the label (squash+merge_me, I suppose?) and then it's going to be merged in 2 days, unless further changes are made.

**Comment by hsyl20** [comment]
@Mikolaj Done! Thanks :)

**Comment by Mikolaj** [comment]
@mergify regresh

**Comment by Mikolaj** [comment]
Mergify is fussy today, so let me prod it manually.

**Comment by Mikolaj** [comment]
@mergify rebase

**Review by jneira — COMMENTED** [review]
late to the party but thanks for working on this (and add the test!)

**Pull request merged**
